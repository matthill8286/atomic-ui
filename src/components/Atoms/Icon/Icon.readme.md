## How to use?
Import the `Icon` component and also import the SVG you want to use directly from the `@excelwithbusiness/webmobile-svg-library` module and nest it. The `Icon` component itself has the Prop `rotate` to allow a rotation of a SVG.

```javascript
import { Icon } from '@/components/Atoms/Icon'
import { IconArrow } from '@excelwithbusiness/webmobile-svg-library'

<Icon>
  <IconArrow />
</Icon>

<Icon rotate={90}> 
  <IconArrow />
</Icon>
```

  &nbsp;

## List of existing Icons

```js

IconAdd

IconArticle

IconBackUi

IconBookmark

IconBookmarkActive

IconCheckmarkCircle

IconCheckmarkCircleOutlined

IconCheckmarkSquare

IconCheckmarkSquareOutlined

IconClear

IconClearCircle

IconClearCircleOutlined

IconClearSquare

IconClearSquareOutlined

IconCross

IconDone

IconFastForward

IconFastRewind

IconFilter

IconForward

IconHeart

IconHeartOutlined

IconHelp

IconHelpOutlined

IconInfo

IconInfoOutined

IconMenuIcon

IconNext

IconPause

IconPlay

IconPlayOutlined

IconSearch

IconSettings

IconSettingsOutlined

IconSkip

IconSkipNext

IconSkipPrevious

IconVolumeDown

IconVolumeLow

IconVolumeMinus

IconVolumeMute

IconVolumeOff

IconVolumePlus

IconVolumeUp

```
