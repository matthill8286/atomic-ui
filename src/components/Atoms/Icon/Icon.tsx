import { FlexAlignItemsOptions } from '@/components/Helper/FlexBox'
import { padding } from '@/styles/sc-shared-functions'
import { css, styled } from '@/styles/styled'
import { PaddingProps, Size, ThemeColors } from '@/types'

export type RotationValues = 0 | 45 | 90 | 180 | 270 | -45 | -90 | -180 | -270

export interface IconProps {
  rotate?: RotationValues
  color?: ThemeColors
  withBorder?: boolean
  padding?: PaddingProps['padding']

  width?: number | keyof Omit<Size, 'xxxs'>
  height?: number | keyof Omit<Size, 'xxxs'>
  animate?: boolean
  display?: string
  alignSelf?: FlexAlignItemsOptions
}

const DefaultSize = '25px'

export const Icon = styled.div<IconProps>(
  ({
    theme,
    color,
    width,
    height,
    rotate = 0,
    animate = false,
    display = 'flex',
    alignSelf,
    withBorder,
  }) => css`
    ${display && `display: ${display};`};
    ${padding};
    ${alignSelf && `align-self: ${alignSelf};`};
    ${withBorder &&
      `border: 2px solid ${theme.color.grey2};
        border-radius: 50%;
        padding: ${theme.spacing.base.xs};`}
    svg {
      width: ${width ? getSizeInPx(width, theme.spacing.base) : DefaultSize};
      height: ${height ? getSizeInPx(height, theme.spacing.base) : DefaultSize};
      transform: rotate(${rotate}deg);
      fill: ${color ? theme.color[color] : 'currentColor'};
      ${animate &&
        css`
          transition: transform 0.3s ease-in-out;
        `};
    }
    svg g,
    svg p {
      fill: ${color ? theme.color[color] : 'currentColor'};
    }
  `
)

const getSizeInPx = (
  size: number | keyof Omit<Size, 'xxxs'>,
  themeSizes: Omit<Size, 'xxxs'>
): string => {
  if (typeof size === 'number') {
    return `${size}px`
  } else {
    return themeSizes[size] || `${DefaultSize}px`
  }
}
