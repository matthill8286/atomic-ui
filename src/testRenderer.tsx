import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
