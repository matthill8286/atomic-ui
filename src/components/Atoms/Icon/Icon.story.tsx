import { storiesOf } from '@storybook/react'
import React from 'react'
import { FlexBox } from '@/components/Helper/FlexBox'
import {
  OtherArrowRightCircle,
  OtherBookmark,
  OtherClose,
  OtherCompleteChecked,
  OtherDone,
  OtherUseful,
  OtherArticle,
  OtherFacebook,
} from '@matthill8286/atomic-icon-library'
import { Icon } from './Icon'
import { Spacer } from '@/components/Atoms/Spacer'

import Readme from './Icon.readme.md'
import { IconButton } from '@/components/Atoms/Button'

storiesOf('Design System/Atoms/Icon', module)
  .add(
    'Icons',
    () => {
      return (
        <>
          <Icon color={'primary'}>
            <OtherArticle />
          </Icon>
          <Spacer size={'xl'} />
          <Icon rotate={90}>
            <OtherCompleteChecked />
          </Icon>
          <Spacer size={'xl'} />
          <Icon height="md" width="md">
            <OtherBookmark />
          </Icon>
          <Spacer size={'xl'} />
          <Icon height="sm" width="sm">
            <OtherClose />
          </Icon>
          <Spacer size={'xl'} />
          <Icon height="lg" width="lg" color={'secondary'}>
            <OtherUseful height="lg" width="lg" />
          </Icon>
          <Spacer size={'xl'} />
          <Icon height={'md'} width={'md'} color={'primary'}>
            <OtherArrowRightCircle />
          </Icon>
        </>
      )
    },
    {
      info: Readme,
    }
  )
  .add(
    'Rotated icon',
    () => {
      return (
        <Icon height={'md'} width={'md'} rotate={90}>
          <OtherArrowRightCircle fill={'primary'} />
        </Icon>
      )
    },
    {
      info: Readme,
    }
  )
  .add('With border', () => {
    return (
      <FlexBox>
        <Icon withBorder width={25} height={25} color="primary">
          <OtherFacebook />
        </Icon>
      </FlexBox>
    )
  })
  .add('Large primary icon button', () => {
    return (
      <FlexBox>
        <IconButton round width={65} height={65} actionType="primary">
          <OtherDone fill={'white'} />
        </IconButton>
      </FlexBox>
    )
  })
