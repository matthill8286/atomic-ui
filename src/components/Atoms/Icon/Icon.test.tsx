import React from 'react'
import { IconArticle } from '@excelwithbusiness/webmobile-svg-library'
import { renderWithTheme } from '@/testRenderer'
import { Icon } from './Icon'

describe('Icon', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(
      <Icon>
        <IconArticle />
      </Icon>
    )
    expect(tree).toMatchSnapshot()
  })
})
