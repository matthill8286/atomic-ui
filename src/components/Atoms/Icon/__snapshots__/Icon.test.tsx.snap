// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 svg {
  width: 25px;
  height: 25px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  fill: currentColor;
}

.c0 svg g,
.c0 svg p {
  fill: currentColor;
}

<div
  className="c0"
>
  <svg
    height={16}
    width={16}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M9.944 0a.74.74 0 01.106.007L9.944 0a.733.733 0 01.442.148l.07.06 4.332 4.267c.025.024.048.05.069.078l-.069-.078a.714.714 0 01.128.17l.011.022a.67.67 0 01.033.077l.007.023a.647.647 0 01.033.21v8.534c0 1.178-.97 2.133-2.167 2.133H4.167C2.97 15.644 2 14.69 2 13.511V2.133C2 .955 2.97 0 4.167 0zm-.722 1.422H4.167a.717.717 0 00-.723.711v11.378c0 .393.324.711.723.711h8.666a.717.717 0 00.723-.71l-.001-7.824h-3.61a.719.719 0 01-.716-.614l-.007-.096V1.422zm2.167 9.245c.399 0 .722.318.722.71a.717.717 0 01-.722.712H5.61a.717.717 0 01-.722-.711c0-.393.323-.711.722-.711zm0-2.845c.399 0 .722.319.722.711a.717.717 0 01-.722.711H5.61a.717.717 0 01-.722-.71c0-.393.323-.712.722-.712zM7.056 4.978c.398 0 .722.318.722.71a.717.717 0 01-.722.712H5.61a.717.717 0 01-.722-.711c0-.393.323-.711.722-.711zm3.61-2.551v1.84l1.867-.001-1.867-1.839z"
      fill="#666"
      fillRule="evenodd"
    />
  </svg>
</div>
`;
